lib/DxcSupport/dxcapi.extval.cpp

```diff
@@ -244,25 +245,27 @@ HRESULT DxcDllExtValidationLoader::CreateInstanceImpl(REFCLSID clsid,
       else {
         // TODO: IDxcCompiler2 not implemented yet, but should
         // error for now
-        hr = E_FAIL;
+        hr = E_UNEXPECTED;
       }
 
       if (FAILED(hr))
         return hr;
 
-      IMalloc *pDefaultMalloc = nullptr;
-      HRESULT hrAlloc = CoGetMalloc(1 /*MEMCTX_TASK*/, &pDefaultMalloc);
-      if (SUCCEEDED(hrAlloc)) {
-        ExternalValidationHelper *evh =
-            Alloc(pDefaultMalloc, DxcCompiler, DxcCompiler3, validatorMajor,
-                  validatorMinor);
-        pDefaultMalloc->Release(); // Alloc holds a copy in m_pMalloc
-        if (!evh)
-          return E_OUTOFMEMORY;
-        hr = evh->QueryInterface(riid, reinterpret_cast<void **>(pResult));
+      CComPtr<IMalloc> pDefaultMalloc;
+      hr = DxcCoGetMalloc(1 /*MEMCTX_TASK*/, &pDefaultMalloc);
+      if (FAILED(hr))
         return hr;
-      }
-      return hrAlloc;
+
+      DxcThreadMalloc TM(pDefaultMalloc);
+
+      ExternalValidationHelper *evh =
+          Alloc(pDefaultMalloc, DxcCompiler, DxcCompiler3, validatorMajor,
+                validatorMinor);
```

Comment:
This is how it is done in dxcapi.cpp, I don't think we're improving things by trying to beat a new path:


```suggestion
      DxcThreadMalloc TM(nullptr);

      ExternalValidationHelper *evh =
          Alloc(DxcGetThreadMallocNoRef(), DxcCompiler, DxcCompiler3, validatorMajor,
                validatorMinor);
```

Similarly for the "2" version of this.

https://github.com/bob80905/directxshadercompiler/blob/add_extvalsupport_class_to_val_infra/tools/clang/tools/dxcompiler/dxcapi.cpp#L110

Reply:
Maybe I'm missing something about why we shouldn't follow the existing pattern here.

Reply:
I don't think we need to change the "2" version, since the function's parameters give access directly to the allocator.

Reply:
As long as you've checked all of the code that's called from here to make sure that it isn't assuming that g_ThreadMallocTls has been set and is indeed explicitly checking the passed in allocator, and that any code added in the future doesn't that same assumption I guess this is fine.